Added @datacommonsorg/client package for fetching csv/json/geojson data.

dwnoble opened the pull request:
Added readme, examples, and tests.

Demo: https://observablehq.com/@datacommonsorg/obesity-vs-sleeping-less-than-7-hours-per-night-in-us-countie

## Usage

Install `@datacommonsorg/client`

```
npm i @datacommonsorg/client
```

Initialize client

```
import { DataCommonsWebClient } from "./DataCommonsClient";

const client = new DataCommonsClient();
```

### Fetch CSV

Fetch total population below the poverty level for each US State

```
const response = await client.getCsv({
  variables: ["Count_Person_BelowPovertyLevelInThePast12Months"],
  parentEntity: "country/USA",
  childType: "State"
});
```

```text
"Count_Person_BelowPovertyLevelInThePast12Months.date","Count_Person_BelowPovertyLevelInThePast12Months.name","Count_Person_BelowPovertyLevelInThePast12Months.unit","Count_Person_BelowPovertyLevelInThePast12Months.unitDisplayName","Count_Person_BelowPovertyLevelInThePast12Months.value","entity.isoCode","entity.name","entity.dcid"
"2021","Population Below Poverty Level Status in Past Year",null,null,769819,"US-AL","Alabama","geoId/01"
"2021","Population Below Poverty Level Status in Past Year",null,null,75016,"US-AK","Alaska","geoId/02"
"2021","Population Below Poverty Level Status in Past Year",null,null,934911,"US-AZ","Arizona","geoId/04"
"2021","Population Below Poverty Level Status in Past Year",null,null,468113,"US-AR","Arkansas","geoId/05"
"2021","Population Below Poverty Level Status in Past Year",null,null,4741175,"US-CA","California","geoId/06"
"2021","Population Below Poverty Level Status in Past Year",null,null,535976,"US-CO","Colorado","geoId/08"
"2021","Population Below Poverty Level Status in Past Year",null,null,351476,"US-CT","Connecticut","geoId/09"
"2021","Population Below Poverty Level Status in Past Year",null,null,109274,"US-DE","Delaware","geoId/10"
"2021","Population Below Poverty Level Status in Past Year",null,null,100618,"US-DC","District of Columbia","geoId/11"
"2021","Population Below Poverty Level Status in Past Year",null,null,2744612,"US-FL","Florida","geoId/12"
"2021","Population Below Poverty Level Status in Past Year",null,null,1441351,"US-GA","Georgia","geoId/13"
"2021","Population Below Poverty Level Status in Past Year",null,null,133740,"US-HI","Hawaii","geoId/15"
...
```

## Web Demo Server

Start web demo server:
```
cd packages/client
npm run dev
# Open [http://localhost:](http://localhost:5173/)
```

## Screenshots from web demo

<img width="1003" alt="init-client" src="https://github.com/datacommonsorg/website/assets/13766/0da652b1-ab11-4d35-bc8b-b0c693c6f59d">
<img width="1003" alt="fetch-data-rows" src="https://github.com/datacommonsorg/website/assets/13766/856650ce-54f2-4ecd-9dff-da9a74034adc">

<img width="1003" alt="google-maps" src="https://github.com/datacommonsorg/website/assets/13766/253a8021-2f93-4cc8-bcbb-3692f0720076">
<img width="998" alt="google-maps-3d" src="https://github.com/datacommonsorg/website/assets/13766/c41c58cf-d63c-4a40-86d6-699795af80a1">

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+      features: dataRows
+        .filter((dataRow) => {
+          const geometryString = dataRow[`entity.${geoJsonProperty}`];
+          if (geometryString !== "string") {
```
curious what would cause the geometryString to come back as "string". Did you mean to check that the `typeof` geometryString is "string"?

dwnoble replied:
Whoops, meant to use `typeof`. Fixed.

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+        : "https://datacommons.org";
+  }
+
+  async getCsv(params: GetDataRowsParams): Promise<string> {
```
add comment doc about the function?

dwnoble replied:
Added

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+    return csvLines.join("\n");
+  }
+
+  async getGeoJSON(params: GetGeoJSONParams): Promise<FeatureCollection> {
```
add comment doc about the function?

dwnoble replied:
Added

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+  async getDataRows(params: GetDataRowsParams): Promise<DataRow[]> {
+    // Fetch variable observations
+    const pointApiResponse =
+      "parentEntity" in params
```
maybe should check `typeof params` instead because param fields could change and this could easily get missed?

dwnoble replied:
I tried a few ways to do this, and the `"parentEntity" in params` was the only thing that seemed to work. `typeof params` always returns the string `"object"`, and `instanceof` doesn't work for types, only classes 🤷 

https://www.turing.com/kb/check-type-of-objects-variables-in-typescript

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+    return Promise.resolve(dataRows);
+  }
+
+  async getFirstNodeValues(params: {
```
add a function comment about what this does? Just reading the name of the function is not super clear

dwnoble replied:
Added, and also commented `getNodePropvals` which might help the clarify things as well

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+      prop: params.prop,
+    });
+    const url = `${this.apiRoot || ""}/api/node/propvals/out?${queryString}`;
+    return (await (await fetch(url)).json()) as ApiNodePropvalOutResponse;
```
nit: maybe split the awaits to two separate lines for easier reading

dwnoble replied:
Done (fixed other instances as well)

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+   * @param entities list of entitites to get data for
+   * @param variables list of variables to get data for
+   * @param date date to get the data for
+   * @param alignedVariables groups of variables that should have the same unit
```
this isn't a param?

dwnoble replied:
Removed

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+
+  /**
+   * Gets and processes the data from /api/observations/point endpoint
+   * @param apiRoot api root
```
this isn't a param?

dwnoble replied:
removed

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+    childType: string;
+    variables: string[];
+    date?: string;
+    facetIds?: string[];
```
include this in the comment doc list of params?

dwnoble replied:
Removed facetIds entirely for now (revisiting in a future PR)

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+  async getObservationsSeries(params: {
+    entities: string[];
+    variables: string[];
+    facetIds?: string[];
```
include this in the list of params in the comment doc? & check the comments for other functions as well

dwnoble replied:
Removed facetIds entirely for now (revisiting in a future PR)

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+        });
+        // Set per-capita data
+        if (TOTAL_POPULATION_VARIABLE in populationObservations.data) {
+          const series =
```
also need to check that entityDcid is in `populationObservations.data[TOTAL_POPULATION_VARIABLE]`

dwnoble replied:
Added!

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+    observations: Observation[],
+    targetDate: string
+  ): Observation | undefined {
+    const index = _.sortedIndexBy(
```
Should either handle empty list for observations and empty string for targetDate or have a comment that the two parameters can't be empty

dwnoble replied:
Good catch- added a handler for when targetDate is blank. It looks like the empty observation case already works, as indexing out of bounds in javascript will return `undefined` which is the return value i want (`[][0] === undefined`)

chejennifer commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
nit: This file is quite large, would be nice if you could split out some functions. Maybe the pure data fetches can go in their own file? May can split out some of the helpers?

dwnoble replied:
Split into two classes - `DataCommonsWebClient` now only makes API calls, and `DataCommonsClient` has the public getCsv/getDataRows/getGeoJSON. I kept the helpers in `DataCommonsClient` for now since the file seems to be a manageable size.

chejennifer reviewed (commented):
Super cool! Mostly just a bunch of nits

dwnoble commented on packages/client/README.md:
````diff
@@ -0,0 +1,191 @@
+```
+import { DataCommonsWebClient } from "./DataCommonsClient";
+
+const client = new DataCommonsClient();
````
This is by design- i'm importing the class on line 24 and instantiating it on line 26 (unless i'm missing something here)

dwnoble commented on packages/client/README.md:
````diff
@@ -0,0 +1,191 @@
+Fetch total population below the poverty level for each US State
+
+```
+const response = await client.getCsv({
````
Done!

dwnoble commented on packages/client/README.md:
````diff
@@ -0,0 +1,191 @@
+```
+[
+  {
+    "entity.dcid": "geoId/01",
````
My intent here was to keep the data structure as flat and CSV-like as possible. This way `getDataRows`, `getCsv`, and `getGeoJSON` all return roughly the same structure. for data rows, users would have to access the data using `response[0]["entity.dcid"]`, which admittedly doesn't look as clean as `response[0].entity.dcid`. Still, i like the consistency of of having the field named "entity.dcid" in all 3 response types

dwnoble commented on packages/client/README.md:
````diff
@@ -0,0 +1,191 @@
+Get total population below the poverty level for each US State
+
+```
+const response = await client.getCsv({
````
Should have been `client.getDataRows` - fixed!

dwnoble commented on packages/client/README.md:
```diff
@@ -0,0 +1,191 @@
+
+### Fetch Data Rows (JSON)
+
+Get total population below the poverty level for each US State
```
Good callout- Per capita can be provided to all 3 `get` methods. Updated the readme & examples to make this more clear.

dwnoble commented on packages/client/src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+) => {
+  const urlSearchParams = new URLSearchParams();
+  Object.keys(obj).forEach((key) => {
+    if (obj[key] === undefined || obj[key] === null) {
```
In this case i do want to include the value if it's equal to `0` which wouldn't be truthy

dwnoble commented on packages/client/src/DataCommonsClient.test.ts:
```diff
@@ -0,0 +1,356 @@
+  });
+};
+
+const buildMockeFetchResponses = (): Record<string, any> => {
```
fixed!

dwnoble commented on packages/client/src/DataCommonsClient.test.ts:
```diff
@@ -0,0 +1,356 @@
+      parentEntity: "country/MOCK",
+      childType: "State",
+    });
+    console.log(JSON.stringify(response, null, 2));
```
removed

dwnoble commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+const DEFAULT_VARIABLE_PROPS = [NAME_ATTRIBUTE];
+
+export interface DatacommonsWebClientParams {
+  /** Web api root endpoint. Default: `"https://datacommons.org/""` */
```
Good catch. fixed!

dwnoble commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+  variables: string[];
+  /** Example: 2023 */
+  date?: string;
+  facetIds?: string[];
```
I took another look at this parameter and couldn't actually get it to do anything on the request. Removing facetIds entirely for now, and ill add back in a future PR

dwnoble commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+  /** Fetch these variable properties from the knowledge graph. Default: `["name"]` */
+  variableProps?: string[];
+  /**
+   * Performs per-capita caluclation for any of these variables.
```
fixed!

dwnoble commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+  constructor(params?: DatacommonsWebClientParams) {
+    const p = params || {};
+    this.apiRoot =
+      p.apiRoot !== undefined
```
Updated

dwnoble commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+      ...params,
+      entityProps: [
+        geoJsonProperty,
+        ...(params.entityProps ? params.entityProps : DEFAULT_ENTITY_PROPS),
```
Updated

dwnoble commented on packages/client/src/types.ts:
```diff
@@ -0,0 +1,145 @@
+  metadata: StatMetadata;
+}
+
+export interface ObservationDate {
```
Removed this and a few other related interfaces

dwnoble commented on packages/client/src/DataCommonsClient.ts:
```diff
@@ -0,0 +1,480 @@
+
+export type NodePropValues = Record<string, Record<string, string | null>>;
+
+class DataCommonsWebClient {
```
Updated file names to use underscore per [ts style guide](http://go/tsjs-style#file-names), and updated class export names to be the camel-case'd file name.

dwnoble commented:
Thank you @chejennifer for the extensive review! I think i've addressed everything. I also bundled into this update:

* Fixed NPM packge.json module definition for publishing on npmjs.com
* Added `turf/rewind` to "rewind" GeoJSON geometries so they show up properly in d3/observable. Not as comprehensive as our website approach, but it's a start

chejennifer commented on packages/client/src/data_commons_client.ts:
```diff
@@ -0,0 +1,352 @@
+  }): Promise<Record<string, string | null>> {
+    const nodePropvals = await this.webClient.getNodePropvals(params);
+    const nodeValues: Record<string, string | null> = {};
+    Object.keys(nodePropvals).forEach((variableDcid) => {
```
nit: nodePropvals aren't keyed by variable dcids, are they? maybe change `variableDcid` to something that describes the keys

dwnoble replied:
Looks like [it is keyed by dcid](https://datacommons.org/api/node/propvals/out?dcids=geoId/18&dcids=Count_Person&prop=name), but it could be any node dcid (variable or entity). Updated to `nodeDcid`

chejennifer reviewed (approved):
thanks for the updates! Just a few more nits & since these will all be new files going into the repo, maybe we can update the copyright to 2024 for all of these files

dwnoble commented on packages/client/src/data_commons_web_client.ts:
```diff
@@ -0,0 +1,185 @@
+
+  /**
+   * Fetches place/entity names
+   * @param entityDcids list of entity DCIDs
```
Fixed

dwnoble commented on packages/client/src/data_commons_client.ts:
```diff
@@ -0,0 +1,352 @@
+const DEFAULT_ENTITY_PROPS = [NAME_ATTRIBUTE, ISO_CODE_ATTRIBUTE];
+const DEFAULT_VARIABLE_PROPS = [NAME_ATTRIBUTE];
+
+export interface DatacommonsClientParams {
```
Fixed

dwnoble commented on packages/client/src/data_commons_client.ts:
```diff
@@ -0,0 +1,352 @@
+   * @returns
+   */
+  async getGeoJSON(params: GetGeoJSONParams): Promise<FeatureCollection> {
+    const geoJsonProperty = params.geoJsonProperty || "geoJsonCoordinatesDP1";
```
done

dwnoble commented on packages/client/src/data_commons_client.ts:
```diff
@@ -0,0 +1,352 @@
+      features: dataRows
+        .filter((dataRow) => {
+          const geometryString = dataRow[`entity.${geoJsonProperty}`];
+          if (typeof geometryString !== "string") {
```
Nice! updated

dwnoble commented on packages/client/src/data_commons_web_client.ts:
```diff
@@ -0,0 +1,185 @@
+
+  /**
+   * Fetches all node property values for the given property name
+   * @param params
```
Done
